### PR TITLE
Keywords

### PR DESCRIPTION
**Pull request by Kink3d**
### Purpose of this PR
Adds user defined Keywords to the graph via the Blackboard. These can be used to wrap graph generated code in ifdefs for performance and compilation reasons.

#### Goals
- Must provide a full featured user keyword system
- Must support predefined keywords such as HDRP’s material quality
- Must be explicit, code readability is a secondary concern
- Must provide minimal interpolators per keyword permutation
- Must allow keyword definition inside Sub Graphs

#### Design
- Create and Edit Keywords on the Blackboard
- Have sections of the Blackboard for Properties and Keywords
- Keywords come in two types (Boolean and Enum), this cannot be changed after creation
- Predefined Keyword definition allows you to switch in the graph based on Keywords you know exist in the pipeline
- Exposing Keywords allows you to set them from the Material inspector
- Dragging Keywords onto the Graph creates Keyword nodes
- Keyword nodes cannot be changed to reference other Keywords (like properties)
- Keyword nodes have one input port per Keyword option
- Keyword nodes update automatically to reflect changes on the Keyword (in the Blackboard)

#### Implementation notes:
- Each node is wrapped in all active permutations individually
   - This is because evaluating the graph per branch in a way that generates code similar to how a human would write the branches is a LOT of work, this method was deemed OK as it was explicit and added less code than writing the full graph per permutation
   - Later we will improve this by evaluating permutations that share lots of code and attempting to patch them together
- Interpolators in HDRP
   - Interpolator structs are if/def in the same manner as the main graph code. This is the ideal for now
   - Packing structs are much more complicated because of how they are made. Therefore for now they written in full, once per permutation.
- Interpolators in URP
   - These are currently not handled. This is because we have a ES customer bug to fix an issue that requires moving URP to use active fields for interpolators (like HDRP). Supporting the current interpolator generation in URP was deemed a waste of time, we will just do the active fields work soon.
- Sub Graphs
   - Keywords inside Sub Graphs are treated the same as regular Keywords.
   - To expose a Keyword from inside a Sub Graph to the Material Inspector you must duplicate it into the main graph and expose it.

#### QA Notes:
- All property refactor work landed in another PR so risk to Properties should be minimal, but still give them a check, especially copying/reordering both Keywords and Properties on the Blackboard.
- Make sure to really test every edge case of deleting/reordering Enum Keyword entries while their nodes on the graph have edges connected. This was prone to failure during development.
- Make sure to check both Generated and Preview code of the graph while testing. Correct scoping of ifdefs in the code is central to the success of this PR.
- This feature has a complicated integration with Sub Graphs. Make sure to test these two features together vigorously.

Full spec: https://favro.com/organization/c564ede4ed3337f7b17986b6/d06b8de3e40da6d340b5b063?card=Uni-86690

---
### Release Notes
See Changelog

---
### Testing status
**Katana Tests**: Coming...

**Manual Tests**: What did you do?
- [ ] Opened test project + Run graphic tests locally
- [ ] Checked new UI names with UX convention
- [x] Tested UI multi-edition + Undo/Redo
- [x] C# and shader warnings (supress shader cache to see them)
- Other: 

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)
Coming...

Any test projects to go with this to help reviewers?
Coming...

---
### Overall Product Risks
**Technical Risk**: None, Low, Medium, High? High - Large feature with lots of moving parts. Large affect on shader generation.

**Halo Effect**: None, Low, Medium, High? Medium - Large risk to shader generation. Most of the risk to properties and blackboard landed in an earlier PR.

---
### Comments to reviewers
See Favro card linked above.

### CODE REVIEW

File: com.unity.render-pipelines.core/Runtime/Utilities/MaterialQuality.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        }
+
+        public static int ToFirstIndex(this MaterialQuality level)
+        {
+            for (var i = 0; i < KeywordNames.Length; ++i)
+            {
+                if ((level & (MaterialQuality) (1 << i)) != 0)
+                    return i;
+            }
+            return -1;
```
Comment (zadi0648) [inline com.unity.render-pipelines.core/Runtime/Utilities/MaterialQuality.cs]: nit-picky, but GetHighestQuality has a line before returning whereas this one does not (I believe this is inconsistent throughout out codebase, but could be nice if consistent within files :p, I think a line after a brace before returning is better)

File: com.unity.render-pipelines.high-definition/Editor/Material/BaseShaderPreprocessor.cs [hunk 2]
```diff
@@ -59,6 +61,22 @@ public BaseShaderPreprocessor()
...
+            if (IsMaterialQualityVariantStripped(hdrpAsset, inputData))
+                return true;
+
+            var shaderMaterialLevel = inputData.shaderKeywordSet.GetMaterialQuality();
+            // if there are material quality defines in this shader
+            if (shaderMaterialLevel != 0
+                // and they don't match the material quality accepted by the hdrp asset
+                && (hdrpAsset.materialQualityLevels & shaderMaterialLevel) == 0)
+                // then strip this variant
+                return true;
```
Comment (zadi0648) [inline com.unity.render-pipelines.high-definition/Editor/Material/BaseShaderPreprocessor.cs]: Base of Peter's comments in other threads, I believe multi-line if statements like this need braces? In this case both the if statement logic and the scope (only because of the comment) are both more than one line.

File: com.unity.shadergraph/Editor/Data/Nodes/AbstractMaterialNode.cs [hunk 3]
```diff
@@ -260,6 +260,10 @@ public virtual void CollectShaderProperties(PropertyCollector properties, Genera
             }
         }
 
+        public virtual void CollectShaderKeywords(KeywordCollector keywords, GenerationMode generationMode)
+        {
+        }
```
Comment (zadi0648) [inline com.unity.shadergraph/Editor/Data/Nodes/AbstractMaterialNode.cs]: I take it that this do-nothing method isn't abstract because it would be annoying to implement the do-nothing method in every single node. Perhaps it would make more sense to take this out of `AbstractMaterialNode` and have it only in `SubGraphNode`, and in the only place it gets called in `GraphUtil` check if the `activeNode is SubGraphNode` and then calling it `as SubGraphNode`?

File: com.unity.shadergraph/Editor/Data/Nodes/Utility/KeywordNode.cs [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+                    List<MaterialSlot> inputSlots = new List<MaterialSlot>();
+                    GetInputSlots(inputSlots);
+
+                    // Store the edges
+                    Dictionary<MaterialSlot, List<IEdge>> edgeDict = new Dictionary<MaterialSlot, List<IEdge>>();
+                    foreach (MaterialSlot slot in inputSlots)
+                        edgeDict.Add(slot, (List<IEdge>)slot.owner.owner.GetEdges(slot.slotReference));
+                    
+                    // Remove old slots
+                    for(int i = 0; i < inputSlots.Count; i++)
```
Comment (zadi0648) [inline com.unity.shadergraph/Editor/Data/Nodes/Utility/KeywordNode.cs]: more nitpickyness; spaces before parentheses seems inconsistent 😝 (imo should always have parentheses before, but perhaps our conventions says otherwise)

File: com.unity.render-pipelines.high-definition/Editor/Material/AxF/AxFShaderPreprocessor.cs [hunk 5]
```diff
@@ -9,6 +9,9 @@ class AxFShaderPreprocessor : BaseShaderPreprocessor
 
         public override bool ShadersStripper(HDRenderPipelineAsset hdrpAsset, Shader shader, ShaderSnippetData snippet, ShaderCompilerData inputData)
         {
+            if (IsMaterialQualityVariantStripped(hdrpAsset, inputData))
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Material/AxF/AxFShaderPreprocessor.cs]: you only need to call IsMaterialQualityVariantStripped in CommonShaderPreprocessor, this will applly to all shader, no need for other call.

File: com.unity.render-pipelines.high-definition/Editor/Material/BaseShaderPreprocessor.cs [hunk 6]
```diff
@@ -59,6 +61,22 @@ public BaseShaderPreprocessor()
             };
         }
 
+        protected static bool IsMaterialQualityVariantStripped(HDRenderPipelineAsset hdrpAsset, ShaderCompilerData inputData)
+        {
+            if (IsMaterialQualityVariantStripped(hdrpAsset, inputData))
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Material/BaseShaderPreprocessor.cs]: you only need to call IsMaterialQualityVariantStripped in CommonShaderPreprocessor, this will applly to all shader, no need for other call.

Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Material/BaseShaderPreprocessor.cs]: I don't understand, isn't this an infinite recursive loop? you call function inside the function ?

File: com.unity.render-pipelines.high-definition/Editor/Material/Lit/LitShaderPreprocessor.cs [hunk 7]
```diff
@@ -10,6 +14,12 @@ class LitShaderPreprocessor : BaseShaderPreprocessor
 
         public override bool ShadersStripper(HDRenderPipelineAsset hdrpAsset, Shader shader, ShaderSnippetData snippet, ShaderCompilerData inputData)
         {
+            if (IsMaterialQualityVariantStripped(hdrpAsset, inputData))
+            {
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Editor/Material/Lit/LitShaderPreprocessor.cs]: you only need to call IsMaterialQualityVariantStripped in CommonShaderPreprocessor, this will applly to all shader, no need for other call.

Also why an exception here? we shouldn't throw exception. is it debgu cocde?

**Comment by sebastienlagarde** [comment]
Poke, any progress on this PR ?

File: com.unity.render-pipelines.high-definition/Editor/Material/BaseShaderPreprocessor.cs (hunk 6 above)
Comment (Kink3d) [reply com.unity.render-pipelines.high-definition/Editor/Material/BaseShaderPreprocessor.cs]: @fredpointzero Can you shed some light here?

File: com.unity.render-pipelines.high-definition/Editor/Material/AxF/AxFShaderPreprocessor.cs (hunk 5 above)
Comment (Kink3d) [reply com.unity.render-pipelines.high-definition/Editor/Material/AxF/AxFShaderPreprocessor.cs]: I have moved this work to a seperate branch `HDRP/material-quality-keywords` as really its a seperate thing.

File: com.unity.render-pipelines.high-definition/Editor/Material/BaseShaderPreprocessor.cs (hunk 6 above)
Comment (Kink3d) [reply com.unity.render-pipelines.high-definition/Editor/Material/BaseShaderPreprocessor.cs]: I have moved this work to a seperate branch `HDRP/material-quality-keywords` as really its a seperate thing.

File: com.unity.render-pipelines.high-definition/Editor/Material/Lit/LitShaderPreprocessor.cs (hunk 7 above)
Comment (Kink3d) [reply com.unity.render-pipelines.high-definition/Editor/Material/Lit/LitShaderPreprocessor.cs]: I have moved this work to a seperate branch `HDRP/material-quality-keywords` as really its a seperate thing.

**Comment by Kink3d** [comment]
@Verasl I've fixed the issues you raised

**Comment by sebastienlagarde** [comment]
Thanks Matt for the change.

@zadi0648 Is there concern left with the PR? Can we move forward. Thanks

**Comment by Kink3d** [comment]
**Test Spin 1:**
HDRP Win Edit: Pass
HDRP Win Play: **Fail**
URP Win Edit: Pass
URP Win Play: Pass
URP OSX Edit: Pass
URP OSX Play: **Fail**
SG Win Edit: Pass
SG Win Play: Pass
SG OSX Edit: Pass
SG OSX Play: **Fail**

**Comment by alindmanUnity** [comment]
> **Test Spin 1:**
> SG OSX Play: **Fail**

SG failure known and present on master, bug filed here to be addressed in a different PR: https://fogbugz.unity3d.com/f/cases/1172471/

**Comment by Kink3d** [comment]
All tests green except OSX Metal SG playmode (known on master)

**Comment by zadi0648** [comment]
https://fogbugz.unity3d.com/f/cases/1176077/ was found today 😕 

We are talking about pushing it through because it _seems_ like this fix will require a trunk change. We got three weeks.
